src/common/lock.ts

```diff
@@ -0,0 +1,18 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+export class Lock {
+    #locked = false;
```

Comment:
How we ensure the thread safety of this `locked` field?

Reply:
Except for the web worker scenario, javascript is in general single thread. For an async function, it will be synchronously executed until an `await` occurs, then it will give the control to the event loop. So the operation
```
if (this.#locked) {
    return; // do nothing
}
this.#locked = true;
```
should be atomic.